# Introduction {.page_break_before}

The regulatory networks that determine cell and tissue identity are robust, yet
remarkably flexible. Transcription factors (TFs) control the expression of genes
by binding to their cognate DNA sequences, TF motifs, in cis-regulatory
elements [@doi:10.1016/j.cell.2018.01.029]. To understand how genetic variation affects binding and to elucidate
the role of TFs in regulatory networks we need to be able to accurately model
binding of TFs to the DNA sequence.

The specificity of DNA-binding proteins can be modeled using various representations [@doi:10.1007/s40484-013-0012-4]. One of the most widely adopted is the position frequency
matrix (PFM). This matrix, a TF motif, contains (normalized) frequencies of each
nucleotide at each position in a collection of aligned binding sites. These PFMs
can be derived from high-throughput experiments such as Chromatin Immunoprecipitation followed by sequencing (ChIP-seq) [@doi:10.1093/nar/gks1221;
@doi:10.1093/nar/gkt1249; @doi:10.1093/nar/gkx1106;
@doi:10.1016/j.molcel.2010.05.004],
HT-SELEX [@doi:10.1101/gr.100552.109] or Protein Binding Microarrays (PBMs) [@doi:10.1093/nar/gku1045]. Through straightforward transformations a PFM can be expressed as a weight matrix, using log likelihoods, or information content, using the Kullback-Leibler divergence.

Here, we present GimmeMotifs, a Python module and set of command-line tools to provide an comprehensive framework for transcription factor motif analysis. Amongst other possibilities it can be used to perfom *de novo* motif analysis, cluster and visualize motifs and to calculate enrichment statistics.
A new ensemble method, *maelstrom*, can be used to determine differential motif activity
between two or more experiments. We illustrate the functionality of GimmeMotifs
using three different examples.
