# Methods

## GimmeMotifs

### Implementation

GimmeMotifs is implemented in Python, with the motif scanning incorporated as a
C module.  The software is developed on GitHub
([https://github.com/simonvh/gimmemotifs/](https://github.com/simonvh/gimmemotifs/))
and documentation is available at
[https://gimmemotifs.readthedocs.io](https://gimmemotifs.readthedocs.io).
Functionality is covered by unit tests, which are run through continuous
integration. GimmeMotifs can be installed via bioconda
[@doi:10.1038/s41592-018-0046-7], see
[https://bioconda.github.io/](https://bioconda.github.io/) for details.  All
releases are also distributed through PyPi [@url:https://pypi.org/] and stably
archived using Zenodo [@url:https://zenodo.org/]. For *de novo* motif search,
14 different external tools are supported (Table 2). All of these are
installed when conda is used for installation. By default,
[genomepy](https://github.com/simonvh/genomepy) is used for genome management
[@doi:10.21105/joss.00320]. In addition, GimmeMotifs uses the following Python
modules: numpy [@url:http://www.numpy.org/], scipy [@url:http://www.scipy.org/],
scikit-learn, scikit-contrib-lightning [@doi:10.1007/s10994-013-5367-2],
seaborn [@doi:10.5281/zenodo.1313201], pysam
[@url:https://github.com/pysam-developers/pysam;@pmid:19505943], xgboost
[@doi:10.1145/2939672.2939785] and pandas. In addition to the
command line tools, all GimmeMotifs functionality is available through a Python
API.

### *De novo* motif prediction pipeline

Originally, GimmeMotifs was developed to predict *de novo* motifs from ChIP-seq
data using an ensemble of motif predictors [@doi:10.1093/bioinformatics/btq636]. The tools
currently supported are listed in Table 2. An input file (BED, FASTA or narrowPeak format) is split into a prediction and
validation set. The prediction set is used to predict motifs, and the validation
set is used to filter for significant motifs. All significant motifs are
clustered to provide a collection of non-redundant *de novo* motifs.  Finally,
significant clustered motifs are reported, along with several statistics to
evaluate motif quality, calculated using the validation set. These evaluation
metrics include ROC AUC, distribution of the motif location relative to the
center of the input (i.e., the ChIP-seq peak summit) and the best match in a
database of known motifs.

| Name          | Version  | Citation |
|:------------- |:---------|:---------|
| [AMD](https://github.com/JiantaoShi/AMD) | Unknown (2012-11-21) | [@doi:10.1371/journal.pone.0024576] |
| [BioProspector](http://ai.stanford.edu/~xsliu/BioProspector/) |  4/15/04 | [@pmid:11262934]  |
| [ChIPMunk](http://autosome.ru/chipmunk/) | V7 10012017 | [@doi:10.1093/bioinformatics/btq488]  |
| [GADEM](https://www.niehs.nih.gov/research/resources/software/biostatistics/gadem/index.cfm) | v1.3.1 | [@doi:10.1089/cmb.2008.16TT]  |
| [HMS](https://dx.doi.org/10.1093/nar/gkp1180) | Unknown (2012-11-23) | [@doi:10.1093/nar/gkp1180] | 
| [Homer](http://homer.ucsd.edu/homer/motif/) | 2 |[ @doi:10.1016/j.molcel.2010.05.004] |
| [Improbizer](https://users.soe.ucsc.edu/~kent/improbizer/improbizer.html) | Unknown (2010-10-01) | [@url:https://users.soe.ucsc.edu/~kent/improbizer/index.html]  |
| [MDmodule](https:/dx.doi.org/10.1038/nbt717) | Unknown (2010-07-29) | [@doi:10.1038/nbt717] |
| [MEME](http://meme-suite.org/) | 4.6.0 | [@pmid:7584402] |
| [MotifSampler](http://bioinformatics.intec.ugent.be/MotifSuite/motifsampler.php) | 3.2 | [@doi:10.1089/10665270252935566] |
| [Posmo](https://dx.doi.org/10.1093/nar/gkr1135) | May-19-2011 | [@doi:10.1093/nar/gkr1135] |
| [Trawler](https://trawler.erc.monash.edu.au/) | 2.0 | [@doi:10.1038/nmeth1061] |
| [Weeder](http://www.beaconlab.it/modtools) | 2.0 | [@doi:10.1002/0471250953.bi0211s47] | 
| [XXmotif](https://github.com/soedinglab/xxmotif) | 1.6 | [@doi:10.1101/gr.139881.112] |

  : **Table 2**: External *de novo* motif prediction tools supported by GimmeMotifs. The version shown is the version string reported by the software. If the version is unknown, the date when the tool was added to the GimmeMotifs repository is shown.

### Motif activity by ensemble learning: maelstrom

GimmeMotifs implements eight different methods to determine differential
enrichment of known motifs between two or more conditions. In addition, these methods can be
combined in a single measure of *motif activity* using rank aggregation. Four
methods work with discrete sets, such as different peak sets or clusters from a
K-means clustering. The hypergeometric test uses motif counts with an empirical
motif-specific FPR of 5%. All other implemented methods use the z-score normalized PFM log-odds
score of the best match. 

To combine different measures of motif significance or activity into a single
score, ranks are assigned for each individual method and combined using rank
aggregation based on order statistics [@doi:10.1038/nbt1203]. This results in a
probability of finding a motif at all observed positions. We use a Python implemention based on the method used in the R package RobustRankAggreg
[@doi:10.1093/bioinformatics/btr709]. The rank aggregation is performed twice,
once with the ranks reversed to generate both positively and negatively
associated motifs.

The hypergeometric test is commonly used to calculate motif enrichment, for instance
by Homer [@doi:10.1016/j.molcel.2010.05.004]. In GimmeMotifs, motifs in each cluster are tested against the
union of all other clusters. The reported value is -log10(p-value) where the p-value
is adjusted by the Benjamini-Hochberg procedure [@url:https://www.jstor.org/stable/2346101]. 

Using the non-parametric Mann-Whitney U test, GimmeMotifs tests the null
hypothesis that the motif log-odds score distributions of two classes are
equal. For each discrete class in the data, such as a cluster, it compares the
score distributions of the class to the score distribution of all other
classes. The value used as activity is the -log10 of the Benjamini-Hochberg
adjusted p-value. 

The two other methods are classification algorithms: random forest using
scikit-learn and a large-scale multiclass classifier using block coordinate
descent [@doi:10.1007/s10994-013-5367-2] as implemented in the
scikit-contrib-lightning module. The classifier in GimmeMotifs uses a l1/l2
penalty with squared hinge loss where the alpha and C parameters are set using
grid search in 10 fold cross-validation procedure.

The other four methods that are implemented relate motif score to an
experimental measure such as ChIP-seq or ATAC-seq signal or expression level.
These are all different forms of regression. In addition to ridge regression,
which  is similar to Motif Activity Response Analysis (MARA)
[@doi:10.1038/ng.375;@doi:10.1101/gr.169508.113], these methods include
regression using boosted trees (XGBoost [@doi:10.1145/2939672.2939785]),
multiclass regression [@doi:10.1007/s10994-013-5367-2] and L1 regularized
regression (LASSO).

## Clustering to create the gimme.vertebrate.v5.0 motif database

We collected all motifs from the following databases: CIS-BP v1.02 [@doi:10.1016/j.cell.2014.08.009], ENCODE [@doi:10.1093/nar/gkt1249], Factorbook [@doi:10.1093/nar/gks1221], HOCOMOCO v11 [@doi:10.1093/nar/gkx1106], HOMER v4.10 [@doi:10.1016/j.molcel.2010.05.004], JASPAR 2018 vertebrates [@doi:10.1093/nar/gkx1126] and SwissRegulon [@doi:10.1093/nar/gkl857]. Motif similarity was calculated using Pearson correlation of motif scores profiles [@doi:10.1186/1471-2105-6-237;@doi:10.7287/peerj.preprints.1302v1] using a sequence that contains each 7-mer or its reverse complement [@doi:10.1093/bioinformatics/btt230]. We then clustered the motifs using agglomerative clustering with complete linkage and connectivity constraints where only motifs with a Pearson r >= 0.5 were considered as neighbors. The number of clusters was set to 1900. After clustering, we discarded all motifs were the sum of the information content of all positions was less than 5. The clustered database, `gimme.vertebrate.v5.0`, is distributed with GimmeMotifs.

## Transcription factor motif database benchmark

We downloaded all hg38 non-ENCODE ChIP-seq peaks from Remap 2018 v1.2 [@doi:10.1093/nar/gkx1092]
([http://tagc.univ-mrs.fr/remap/index.php?page=download](http://tagc.univ-mrs.fr/remap/index.php?page=download)).
We removed all factors with fewer than 1000 peaks and created regions of 100 bp
centered at the peak summit. Background files were created for each peak set
using bedtools shuffle [@doi:10.1093/bioinformatics/btq033], excluding the hg38 gaps and the peak regions.
The ROC AUC and Recall at 10% FDR statistics were calculated using `gimme roc`.
The motif databases included in the comparison are listed in Table 3. We only included public databases that can be freely accessed and downloaded.

| Name          | Version  | Citation | 
|:------------- |:---------|:---------|
| [Factorbook](http://www.factorbook.org/) | Sep. 2012 | [@doi:10.1093/nar/gks1221] 
| [SwissRegulon](http://swissregulon.unibas.ch/sr/swissregulon) | Nov. 2006 | [@doi:10.1093/nar/gkl857] |
| [GimmeMotifs vertebrate](https://gimmemotifs.readthedocs.io) | v5.0 | | 
| [HOCOMOCO](http://hocomoco11.autosome.ru/) | v11 | [@doi:10.1093/nar/gkx1106]
| [Homer](http://homer.ucsd.edu/homer/) | v4.10 | [ @doi:10.1016/j.molcel.2010.05.004] |
| [JASPAR](https://jaspar.genereg.net/) | 2018 | [@doi:10.1093/nar/gkx1126] |
| [ENCODE](http://compbio.mit.edu/encode-motifs/) | Dec. 2013 | [@doi:10.1093/nar/gkt1249]
| [Madsen](http://bioinformatik.sdu.dk/solexa/webshare/IMAGE/IMAGE_v1.1.tar.gz) | 1.1 | [@doi:10.1101/gr.227231.117]
| [RSAT vertebrate clusters](http://pedagogix-tagc.univ-mrs.fr/rsat/data/published_data/Castro_2016_matrix-clustering/) | Sep. 2017 | [@doi:10.1093/nar/gkx314] 
 
  : **Table 3**: Motif databases.

The workflow is implemented in snakemake [@doi:10.1093/bioinformatics/bts480] and is available at
[https://github.com/vanheeringen-lab/gimme-analysis](https://github.com/vanheeringen-lab/gimme-analysis).

## `De novo` motif prediction benchmark

We downloaded all spp ENCODE peaks (January 2011 data freeze) from the EBI FTP
([http://ftp.ebi.ac.uk/pub/databases/ensembl/encode/integration_data_jan2011/byDataType/peaks/jan2011/spp/optimal/](http://ftp.ebi.ac.uk/pub/databases/ensembl/encode/integration_data_jan2011/byDataType/peaks/jan2011/spp/optimal/)).
We selected the top 5000 peaks and created 100bp regions centered on the peak
summit. As background we selected 100 bp regions flanking the original peaks.
For the `de novo` motif search default settings for `gimme motifs` were used. The
workflow is implemented in snakemake [@doi:10.1093/bioinformatics/bts480] and is available at
[https://github.com/vanheeringen-lab/gimme-analysis](https://github.com/vanheeringen-lab/gimme-analysis).

## Motif analysis of hematopoietic enhancers

To illustrate the functionality of `gimme maelstrom` we analyzed an integrated
collection of hematopoietic enhancers. We downloaded all H3K27ac ChIP-seq and
DNase I data from BLUEPRINT and hematopoietic DNase I
data from ROADMAP (Supplementary Table S1). All DNase I data were processed using
the Kundaje lab DNase pipeline version 0.3.0
[https://github.com/kundajelab/atac_dnase_pipelines](https://github.com/kundajelab/atac_dnase_pipelines)
[@doi:10.5281/zenodo.156534].  The ChIP-seq samples were processed using the
Kundaje lab AQUAS TF and histone ChIP-seq pipeline
[https://github.com/kundajelab/chipseq_pipeline](https://github.com/kundajelab/chipseq_pipeline).
For all experiments from BLUEPRINT we used the aligned reads provided by EBI.
All ROADMAP samples were aligned using bowtie2 [@doi:10.1038/nmeth.1923] to the
hg38 genome. DNase I peaks were called using MACS2 
[@doi:10.1186/gb-2008-9-9-r137] according to the default settings of the DNase pipeline. We merged all DNase I peak files and centered
each merged peak on the summit of the strongest individual peak. H3K27ac reads
were counted in a region of 2kb centered at the summit (Supplementary Table S2)
and read counts were log2-transformed and scaled. We removed all samples that
were treated and averaged all samples from the same cell type. We then selected
all enhancers with at least one sample with a scaled log2 read count of 2,
sorted by the maximum difference in normalized signal between samples and
selected the 50,000 enhancers with the largest difference. Using this enhancer
collection as input, we ran `gimme maelstrom` using default settings. The motif
analysis workflow is implemented in a Jupyter notebook and is available at
[https://github.com/vanheeringen-lab/gimme-analysis](https://github.com/vanheeringen-lab/gimme-analysis).
