# Methods

## GimmeMotifs

### Implementation

GimmeMotifs is implemented in Python, with the motif scanning incorporated as a
C module.  The software is developed on GitHub
([https://github.com/simonvh/gimmemotifs/](https://github.com/simonvh/gimmemotifs/))
and documentation is available at
[https://gimmemotifs.readthedocs.io](https://gimmemotifs.readthedocs.io).
Functionality is covered by unit tests, which are run through continuous
integration.  GimmeMotifs can be installed via bioconda [@bioconda], see
[https://bioconda.github.io/](https://bioconda.github.io/) for details.  All
releases are also distributed through PyPi and stably archived using Zenodo
[ref].  For *de novo* motif search, [14] different external tools are supported
(Table [X]). All of these are installed when conda is used for installation. By
default, [genomepy](https://github.com/simonvh/genomepy) is used for genome
management[@ref genomepy]. In addition, GimmeMotifs uses the following Python modules: numpy,
scipy, scikit-learn, scikit-contrib-lightning, seaborn
[https://doi.org/10.5281/zenodo.1313201], pysam, xgboost, pandas [see
https://www.scipy.org/citing.html for citations].  In addition to the command
line tools, all GimmeMotifs functionality is available through a Python API.

### *De novo* motif prediction pipeline

Originally, GimmeMotifs was developed to predict *de novo* motifs from ChIP-seq
data using an ensemble of motif predictors [@vanHeeringen2009]. The tools
currently supported are listed in Table [X]. The pipeline is depicted in [Fig
X]. An input file (BED, FASTA or narrowPeak format) is split in a prediction and
validation set. The prediction set is used to predict motifs, and the validation
set is used to filter for significant motifs. All significant motifs are
clustered to provide a collection of non-redundant *de novo* motifs.  Finally,
significant clustered motifs are reported, along with several statistics to
evaluate motif quality, calculated using the validation set. These evaluation
metrics include ROC AUC, distribution of the motif location relative to the
center of the input (i.e., the ChIP-seq peak summit) and the best match in a
database of known motifs.

| Name          | Citation |
|:------------- |:---------|
| [Homer](http://homer.ucsd.edu/homer/motif/) | [ @Heinz2010] |
| [Trawler](https://trawler.erc.monash.edu.au/) | [@Ettwiller2007] |
| [GADEM](https://www.niehs.nih.gov/research/resources/software/biostatistics/gadem/index.cfm) | [@Li2009]  |
| [MDmodule](@Liu2002) |
| [MEME](http://meme-suite.org/) | [@Bailey1994] |
| [MotifSampler](http://bioinformatics.intec.ugent.be/MotifSuite/motifsampler.php) | [@Thijs2002] |
| [Improbizer](https://users.soe.ucsc.edu/~kent/improbizer/improbizer.html) |   |
| [BioProspector](http://ai.stanford.edu/~xsliu/BioProspector/) |   |
| [Posmo](https://dx.doi.org/10.1093/nar/gkr1135) |   |
| [ChIPMunk](http://autosome.ru/chipmunk/) |   |
| [AMD](https://github.com/JiantaoShi/AMD) |  |
| [HMS](https://doi.org/10.1093/nar/gkp1180) | | 
| [XXmotif](https://github.com/soedinglab/xxmotif) |  |

  : Table [X]: External *de novo* motif prediction tools supported by GimmeMotifs.


### Motif activity by ensemble learning: maelstrom

GimmeMotifs implements eight different methods to determine differential motif
enrichment between two or more conditions. In addition, these methods can be
combined in a single measure of *motif activity* using rank aggregation. Four
methods work with discrete sets, such as different peak sets or clusters from a
K-means clustering. The hypergeometric test uses motif counts with an empirical
motif-specific FPR of 5%. All other implemented methods use the PWM log-odds
score of the best match. 

The hypergeometric test is commonly used, for instance
by Homer [@Homer]. In GimmeMotifs, motifs in each cluster are tested against the
union of all other clusters. The reported value is -log10(p-value) where the p-value
adjusted by the Benjamini-Hochberg procedure [@ref BH]. 

Using the non-parametric
Mann-Whitney U test, GimmeMotifs tests the null hypothesis that that the motif
log-odds score distributions of two classes are equal. For each discrete class
in the data, such as a cluster, it compares the score distributions of the class
to the score distribution of all other classes. The value used as activity is
the -log10 of the Benjamini-Hochberg adjusted p-value. 

The two other methods are classification algorithms: random forest
using scikit-learn and a large-scale multiclass classifier using
block coordinate descent [@Blondel2013] as implemented in the 
scikit-contrib-lightning module. The classifier in GimmeMotifs uses a l1/l2 penalty
with squared hinge loss where the alpha and C parameters are set using grid
search in 10 fold cross-validation procedure.

The other four methods that are implemented relate motif score to an
experimental measure such as ChIP-seq or ATAC-seq signal or expression level.
These are all different forms of regression. In addition to ridge regression, which  is similar to
Motif Activity Response Analysis (MARA) [@mara_nimwegen], these methods include
regression using boosted trees (XGBoost), multiclass regression [@Blondel2013]
and L1 regularized regression (LASSO).

To combine different measures of motif significance or activity into a single
score, ranks are assigned for each individual method and combined using rank
aggregation based on order statistics [@Aerts2006]. This results in a
probability of finding a motif at all observed postiions. We use the method
implemented in the R package RobustRankAggreg [@Kolde2012]. The rank aggregation
is performed twice, once with the ranks reversed to generate both positively and
negatively associated motifs.

## Clustering

## Transcription factor motif database benchmark

We downloaded all ChIP-seq peaks from Remap 2018 v1.2 [@Cheneby2018]
([http://tagc.univ-mrs.fr/remap/index.php?page=download](http://tagc.univ-mrs.fr/remap/index.php?page=download)).
We removed all factors with fewer than 1000 peaks and created regions of 100 bp
centered at the peak summit. Background files were created for each peak set
using bedtools shuffle [@Quinlan], excluding the hg19 gaps and the peak regions.
The ROC AUC and Recall at 10% FDR statistics were calculated using gimme roc.
The workflow is implemented in snakemake [@Koster2012] and can be downloaded at
[https://github.com/simonvh/gimme_analysis](https://github.com/simonvh/gimme_analysis).


## De novo motif benchmark

We downloaded all spp ENCODE peaks (January 2011 data freeze) from the EBI FTP
([http://ftp.ebi.ac.uk/pub/databases/ensembl/encode/integration_data_jan2011/byDataType/peaks/jan2011/spp/optimal/](http://ftp.ebi.ac.uk/pub/databases/ensembl/encode/integration_data_jan2011/byDataType/peaks/jan2011/spp/optimal/)).
We selected the top 5000 peaks and created 100bp regions centered on the peak
summit. As background we selected 100 bp regions flanking the original peaks.
For the the novo motif search default settings for gimme motifs were used. The
workflow is implemented in snakemake [@Koster2012] and can be downloaded at
[https://github.com/simonvh/gimme_analysis](https://github.com/simonvh/gimme_analysis).

## Motif analysis of hematopoietic enhancers

To illustrate the functionality of gimme maelstrom, we analyzed an integrated data set hematopoietic
enhancers. We downloaded all H3K27ac ChIP-seq and DNase I data from BLUEPRINT
(Supplemental Table SX) and hematopoietic DNase I data from ROADMAP (Supplemental Table SX). All DNase I data were processed using the Kundaje lab DNase pipeline version 0.3.0 [https://github.com/kundajelab/atac_dnase_pipelines](https://github.com/kundajelab/atac_dnase_pipelines) [@ref https://zenodo.org/record/156534]. 
The ChIP-seq samples were processed using the Kundaje lab AQUAS TF and histone ChIP-seq pipeline [https://github.com/kundajelab/chipseq_pipeline](https://github.com/kundajelab/chipseq_pipeline). 
For all experiments from BLUEPRINT we used the aligned reads provided by EBI. All ROADMAP samples were aligned using bowtie2 [@ref] to the hg38 genome. DNase I peaks were called using MACS2 [@ref]. 
